### PR TITLE
Decoupling eye tracking data query from eye based gaze selection

### PR DESCRIPTION
**Pull request by vaoliva**
## Overview

The HandInteractionExamples scene aims to have behavior that matches the hololens 2 shell when used with the hololens 2 profile. However, when eye tracking is enabled (either in editor or on device), two issues occur:

The cursor in HandInteractionExamples follows the eyes instead of the head
When a hand is present, the eye cursor will apppear whenever a hand is near a grabbable or touchable object

## To reproduce

Enable "simulate eye position" in MRTK input simulation profile
Run HandInteractionExample scene
Bring hand near grabbable
Notice that the cursor dot appears

## Expected behavior

Cursor remains invisible and the cursor follows the head, not the eyes.
Eye tracking ray can be accessed via InputRayUtils.TryGetEyeRay
If the user wishes to have the cursor follow the gaze direction, he/she can enable this by having the gazeprovider use the eye gaze data, but by default it will use the head data.

Related to #6496

## Changes

- Enables eye ray query if eye tracking data available even if gaze not eye-based 
- Matches eye based and head gaze cursor state behavior

- Fixes: #6541 
- Fixes: #7220

### CODE REVIEW

**Comment by julenka** [comment]
I believe this is also related to #7220

**Comment by sostel** [comment]
If the default behavior is to have head gaze enabled, you could simply remove the Mixed Reality Eye Tracking provider from the MRTK DefaultHoloLens2 profile instead of changing every single scene? 

I was not aware that the HandTrackingInteraction scene is necessarily aiming at exactly mimicking the Shell. If that's so, it should be communicated better to developers.

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs [hunk 1]
```diff
@@ -51,28 +51,6 @@ public void UpdateState(int numNearPointersActive, int numFarPointersActive, int
                 eyeGazeValid = isEyeGazeValid;
             }
 
-            if (isEyeGazeValid)
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: Interesting, @wiwei and @sostel I remember you added this separate state machine for eye gaze to fix a specific bug, regarding eye tracking -- that hand interactions were not working with eyes. See #4482. With this change, can you still do near manipulations with eye tracking? Does this still pass the tests that will had added to verify that hands work with eyes?

Comment (wiwei) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: So it's been quite a while since I did that but from what I remember if eye gaze was enabled, if you had any hands in the scene that had active far/near pointers, it would disable eye gaze, and you had to say "select" to bring back eye gaze. It was essentially unusable from a user perspective.

It feels like this change seems to rip that out and bring it back to the old problem, but I also don't know if maybe the change to TryGetEyeGazeRay above is actually want makes it all work out. So maybe that actually is the better/right fix since it would enable gaze raycasting even if the eye gaze state isn't in the "active" state.

If https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4431 isn't an issue with this code, then this change should be fine, though FWIW yeah there were some tests added to validate behavior, which should be changed to account for how cursors aren't visible now (but the ray itself should still be valid, etc)

Comment (vaoliva) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: TryGetEyeGazeRay is trying to provide eye gaze data even if gaze is not eye based, but if that is the case hit position is not being handled, so all user have is ray direction. 
[Documentation](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/EyeTracking/EyeTracking_TargetSelection.html#example-4-use-hand-rays-and-eye-gaze-input-together) says it is possible to use hands and eye gaze at the same time. Should this be achieved through the TryGetEyeGazeRay instead?

**Comment by julenka** [comment]
>I was not aware that the HandTrackingInteraction scene is necessarily aiming at exactly mimicking the Shell. If that's so, it should be communicated better to developers.

@cre8ivepark to confirm, but my understanding is that HandInteracctionExamples scene should demonstrate behaviors that match shell parity. @sostel for communicating better to developers, would you recommend putting a note in MRTK getting started like "for an example of a scene that demonstrates interactions that are similar to the HoloLens shell, see the HandInteractionExamples scene"? We could put that in.

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: > Should this be achieved through the TryGetEyeGazeRay instead?

Sorry but I'm not understanding what you mean here, could you explain a bit more?

File: Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs [hunk 2]
```diff
@@ -163,6 +163,15 @@ public IMixedRealityInputSource GazeInputSource
...
 
         private Ray latestEyeGaze = default(Ray);
+        /// <summary>
+        /// The most recent eye tracking ray
+        /// </summary>
+        public Ray LatestEyeGaze 
+        { 
+            get { return latestEyeGaze; } 
+            private set { latestEyeGaze = value; } 
+        }
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs]: nit: use get => latestEyeGaze

File: Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs [hunk 3]
```diff
@@ -228,8 +237,8 @@ public override void OnPreSceneQuery()
                 if (gazeProvider.useEyeTracking && gazeProvider.IsEyeTrackingAvailable)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs]: isn't this gazeProvider.IsEyeGazevalid?

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs [hunk 4]
```diff
@@ -16,11 +16,21 @@ public interface IMixedRealityEyeGazeProvider : IMixedRealityGazeProvider
         /// </summary>
         bool IsEyeGazeValid { get; }
 
+        /// <summary>
+        /// Whether eye tracking data is available.
+        /// </summary>
+        bool IsEyeTrackingAvailable { get; }
+
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs]: I'm not exactly understanding the difference between IsEyeGazeValid  and IsEyeTrackingAvailable?
IsEyeTrackingAvailable also sounds more like is the capability available compared to is recent data available/valid

Comment (vaoliva) [reply Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs]: IsEyeGazeValid needs both eye tracking to be available and eye tracking selected as default gaze source. IsEyeTrackingAvailable is there if user wants to query eye tracking ray, even if head gaze is the default. 

Perhaps IsEyeGazeValid should be called something like IsGazeEyeBased ?

File: Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs (hunk 3 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs]: Answered below.

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: What I meant is that when eye is not selected as default gaze, if eye tracking data is available, the user can query the eye ray but If user wants to get focus information about that ray this is not implemented. 

Now I understand that this should be handled in a separate PR if needed.

File: Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs (hunk 2 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazeProvider.cs]: changed

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs (hunk 4 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs]: > Perhaps IsEyeGazeValid should be called something like IsGazeEyeBased ?

IsGazeEyeBased makes more sense to me, or IsGazeInputEyeBased

**Comment by julenka** [comment]
> Proper loading of scenes linked to MRTKExamplesHub scene still needs to be redefined, as currently the scenes are loaded using additive mode using hub scene configuration and the same configuration of eye gaze is shared between all scenes.

FYI @cre8ivepark. If we don't have a bug to track this, can somebody please create one?

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs [hunk 5]
```diff
@@ -70,6 +70,50 @@ public IEnumerator TestGazeCursorArticulated()
...
+            // No hands, default cursor should be visible
+            Assert.IsTrue(CoreServices.InputSystem.GazeProvider.GazeCursor.IsVisible, "Head gaze cursor should be visible");
+
+            // Hand up near grabbable
+            TestHand hand = new TestHand(Handedness.Right);
+            yield return hand.Show(Vector3.forward * 4.05f);
+            yield return null;
+
+            var handInput = inputSystem.DetectedInputSources.Where(x => x.SourceName.Equals("Right Hand")).First();
+            IMixedRealityNearPointer grabPointer = (IMixedRealityNearPointer)handInput.Pointers.Where(x => x.PointerName.Contains("Grab")).First();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: FYI, you can use PointerUtils.GetPointer<GrabPointer>() to simplify this lookup (no need for handInput and the LINQ query)

**Comment by thalbern** [comment]
LGTM - but i would like @sostel to do a final sign off on this if possible

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs (hunk 5 above)
Comment (sostel) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: I don't like renaming "IsEyeGazeIsValid" - Eye gaze is also used for other things than input (heatmaps). It's important to know when eye gaze is valid and when not. For example, when a user is blinking, the eye gaze may be invalid, which doesn't mean it's in general not meant to be used as an input.

**Comment by sostel** [comment]
> > I was not aware that the HandTrackingInteraction scene is necessarily aiming at exactly mimicking the Shell. If that's so, it should be communicated better to developers.
> 
> @cre8ivepark to confirm, but my understanding is that HandInteracctionExamples scene should demonstrate behaviors that match shell parity. @sostel for communicating better to developers, would you recommend putting a note in MRTK getting started like "for an example of a scene that demonstrates interactions that are similar to the HoloLens shell, see the HandInteractionExamples scene"? We could put that in.

Yes, some info in the scene itself would be nice.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs (hunk 5 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: I'm confused -- I thought according to Vanessa IsEyeGazeValid means "Is Eye Gaze Being used to control Input", not "is eye gaze information usable / valid currently". 

Perhaps there could be a variable IsEyeGazeValid which means IsEyeTrackingAvailabe && !IsUserBlinking, and a second variable IsEyeGazeUsedToControlnput.

I think the main concern Vanessa & I have is "IsEyeGazeValid" actually seems to mean "is eye tracking currently being used to controle the GazeProvider in MRTK". At least that's how I understand it.

Comment (sostel) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: Hm, yes, I understand. If I remember correctly, IsEyeGazeValid is not related to whether it is used as input or not. 

Here is some more documentation:
https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/EyeTracking/EyeTracking_EyeGazeProvider.html

_**IsEyeGazeValid**: Indicates whether the current eye tracking data is valid. It may be invalid due to exceeded timeout (should be robust to the user blinking though) or lack of tracking hardware or permissions. Check out our _**Missing eye calibration notification**_ sample that explains how to detect whether a user is eye calibrated and to show an appropriate notification._

In addition, even if the interaction doesn't use eye tracking as an input, developers may still want to access eye tracking data for attention analysis.

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs [hunk 6]
```diff
@@ -64,13 +64,13 @@ public override object[] GetData(string inputType, string inputStatus, EyeTracki
                 0,
 
                 // Smoothed eye gaze signal 
-                EyeTrackingProvider.IsEyeGazeValid ? EyeTrackingProvider.GazeOrigin.x : 0,
-                EyeTrackingProvider.IsEyeGazeValid ? EyeTrackingProvider.GazeOrigin.y : 0,
-                EyeTrackingProvider.IsEyeGazeValid ? EyeTrackingProvider.GazeOrigin.z : 0,
+                EyeTrackingProvider.IsGazeInputEyeBased ? EyeTrackingProvider.GazeOrigin.x : 0,
```
Comment (sostel) [inline Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs]: The LogStructureEyeGaze.cs is used as an example for tracking visual attention. Ideally, this should be possible even if eye gaze is not used as an input.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs (hunk 5 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: It seems like this has changed, since [according to Vanessa](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7246#discussion_r377613528):

> IsEyeGazeValid needs both eye tracking to be available and eye tracking selected as default gaze source. IsEyeTrackingAvailable is there if user wants to query eye tracking ray, even if head gaze is the default.

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs [hunk 7]
```diff
@@ -23,6 +23,11 @@ private void Update()
             var eyeGazeProvider = CoreServices.InputSystem?.EyeGazeProvider;
             if (eyeGazeProvider != null)
             {
+                if(!eyeGazeProvider.UseEyeTracking)
+                {
+                    eyeGazeProvider.UseEyeTracking = true;
```
Comment (sostel) [inline Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs]: Why is this set to true in the code? And why in the Update function? Maybe you want to switch between different head and eye gaze modes at run time?

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs (hunk 5 above)
Comment (sostel) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: That's a bug with mingling head and eye gaze together in GazeProvider. It would probably be better to handle them as independent input sources. I don't think that this was thoroughly tested and understood when Kurtis originally set it up. Nevertheless, there is a difference between whether the eye tracking data is valid and whether it's used as an input. There could be scenarios in which I want to use the head gaze as a pointer and still be able to access eye tracking data - and in that case it would be great to understand whether the data is valid or not.

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs (hunk 6 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs]: This change only reflects field name change from IsGazeValid to IsGazeInputEyeBased . 
IsGazeInputEyeBased still returns UseEyeTracking && IsEyeTrackingAvailable.

So I believe this should be implemented in a separate PR?

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs (hunk 7 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs]: The scene uses the script LoadProfilesOnStartup.cs , this script resets the profile on start. 

Calling the profile initialization on runtime is instantiating a new GazeProvider.cs on the Camera object, and even if we configure GazeProvider to have UseEyeTracking = true on the parent scene, the new GazeProvider has the default value of UseEyeTracking = false.  

This is the reason having UseEyeTracking = true as default was apparently solving the EyeTracking Demo scenes issues.

Solutions for this issue include not changing profile on startup, changing UseEyeTracking on code, or implementing a better way of Profile change which is probably on the way.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs (hunk 5 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/FocusProviderTests.cs]: Before PR, IsEyeGazeValid and IsEyeTrackingValid were calculated like that:

IsEyeTrackingAvailable  returns true if there was recent eye tracking data received from device
IsEyeGazeValid returns IsEyeTrackingAvailable && UseEyeTracking

Documentation says that IsEyeGazeValid reflects valid eye tracking data, but this is not actiually true, as it returns false if head gaze is selected. IsEyeTrackingAvailable actually reflects valid eye data.

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs (hunk 6 above)
Comment (sostel) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs]: The choice of name is confusing and reflects that whether eye gaze data is available is not the same as setting whether eye gaze is used as an input control.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs [hunk 8]
```diff
@@ -158,7 +158,7 @@ private void Update()
             // Try to manually poll the eye tracking data
             if (eyeGazeProvider != null 
                 && eyeGazeProvider.UseEyeTracking
```
Comment (sostel) [inline Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs]: I find the difference between "UseEyeTracking" and "IsGazeInputEyeBased" confusing. Shouldn't "UseEyeTracking" be used as an indicator to whether eye-gaze input is used in contrast to "IsGazeInputEyeBased"?

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/InputRayUtils.cs [hunk 9]
```diff
@@ -33,11 +33,18 @@ public static bool TryGetEyeGazeRay(out Ray ray)
...
-            ray.origin = eyeGazeProvider.GazeOrigin;
-            ray.direction = eyeGazeProvider.GazeDirection;
+            if (eyeGazeProvider.UseEyeTracking)
+            {
+                ray.origin = eyeGazeProvider.GazeOrigin;
+                ray.direction = eyeGazeProvider.GazeDirection;
+            }
+            else
+            {
+                ray = eyeGazeProvider.LatestEyeGaze;
```
Comment (sostel) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/InputRayUtils.cs]: I do not understand why this was added. How is "eyeGazeProvider.GazeOrigin" and "...GazeDirection" different to "eyeGazeProvider.LatestEyeGaze"?

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (sostel) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: If I read it correctly, you removed handling automatically hiding the eye gaze cursor when the hand comes near a target. Where or how is this handled now? 
Also, why do you feel that ensuring a correctly returned eye gaze ray even if not set as primary input should be handled in a separate PR? I feel that it would fall under decoupling eye tracking data from eye-based control.

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs [hunk 10]
```diff
@@ -12,15 +12,25 @@ namespace Microsoft.MixedReality.Toolkit.Input
...
-        /// Whether eye gaze is valid. It may be invalid due to timeout or lack of tracking hardware or permissions.
+        /// Whether eye tracking data is currently been used for gaze rather then head pose.
         /// </summary>
-        bool IsEyeGazeValid { get; }
+        bool IsGazeInputEyeBased { get; }
+
+        /// <summary>
+        /// Whether eye tracking data is available. It may be unavailable due to timeout or lack of tracking hardware or permissions.
+        /// </summary>
+        bool IsEyeTrackingAvailable { get; }
```
Comment (sostel) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityEyeGazeProvider.cs]: Confusing. Why did you rename "IsEyeGazeValid" to "IsEyeTrackingAvailable"? Eye tracking data may be available, but not valid (e.g., during blinks).

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md [hunk 11]
```diff
@@ -11,13 +11,9 @@ If you are new to eye tracking on HoloLens 2 and to how eye tracking is set up i
 We will go into detail on how to address each of them further below.
 
 1. An _'Eye Gaze Data Provider'_ must be added to the input system. This provides eye tracking data from the platform.
-2. The GazeProvider must have its _'Use Eye Tracking'_ property set to true. Note that true is the default value (so no special
-   action is required unless you have actively unchecked this property.)
```
Comment (sostel) [inline Documentation/EyeTracking/EyeTracking_BasicSetup.md]: Why did you remove this? This is part of the basic instructions to get eye tracking input working in MRTK. It's meant as a short overview check list before going into more detail below. Especially now that by default Eye Tracking is unchecked in most scenes, it's even more important to point this out so that devs can better understand why eye tracking is not working for them.

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md [hunk 12]
```diff
@@ -11,13 +11,9 @@ If you are new to eye tracking on HoloLens 2 and to how eye tracking is set up i
...
-3. The _'GazeInput'_ capability must be enabled in the application manifest.
+2. The _'GazeInput'_ capability must be enabled in the application manifest.
    **Currently this is only available in Visual Studio and through the MRTK build tool**
-4. The HoloLens **must** be eye calibrated for the current user. Check out our [sample for detecting whether a user is eye calibrated or not](EyeTracking_IsUserCalibrated.md).
-
-**IMPORTANT:** If any of the above requirements are not met, the application will automatically fall back to head-based gaze tracking.
```
Comment (sostel) [inline Documentation/EyeTracking/EyeTracking_BasicSetup.md]: Why did you remove this note?

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md [hunk 13]
```diff
@@ -57,15 +53,6 @@ You can simply select _DefaultMixedRealityToolkitConfigurationProfile_ and then
 
 ![MRTK](../Images/EyeTracking/mrtk_setup_eyes_dataprovider.jpg)
 
-### Enabling eye tracking in the GazeProvider
-
```
Comment (sostel) [inline Documentation/EyeTracking/EyeTracking_BasicSetup.md]: Why did you move this section?

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: > If I read it correctly, you removed handling automatically hiding the eye gaze cursor when > the hand comes near a target. Where or how is this handled now?

Now the logic for hiding gaze cursor when hand is visible is the same for both head or eye based. Hides when hand visible, unhides with "select" speech command. This is handled by FocusProvider and GazePointerVisibilityStateMachine.

> Also, why do you feel that ensuring a correctly returned eye gaze ray even if not set as primary input should be handled in a separate PR? I feel that it would fall under decoupling eye tracking data from eye-based control.

If eye is not set as gaze input, the eye gaze Ray can be queried from TryGetEyeGazeRay. If available, it will return the correct origin and direction of the eye ray. What cannot be queried is information about what object have eye focus, as FocusProvider is only handling gaze pointer from.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs (hunk 8 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs]: Agree that the names are confusing, although having useEyeTracking does not guarantee that eye is input for gaze pointer.

**Comment by davidkline-ms** [comment]
@vaoliva, @sostel had some recent questions. Once they are resolved, it looks like this should be good for merging.

**Comment by thalbern** [comment]
I've added the do not merge label for now until we resolve / address sophies concerns

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: This looks correct to me - We would want the same cursor behavior for eye gaze and head gaze, no? 
Also the data for querying the eye ray is available (InputRayUtils.TryGetEyeGaze) - the only thing that's missing is that the focus provider won't be able the tell you which object is focused with the eye ray in case the head gaze is active as we only have the target from the gaze pointer - we can add a follow up task for being able to query the object that is hit with the eye tracking ray in case we're in head gaze mode to have both of them available - but it's not something you would need eg for creating a heat map - for that the ray should be enough

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs (hunk 6 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs]: @sostel this was only a rename to fix the confusion that happened in gaze provider with overriding the IsEyeGazeValid flag but actually turning it into something else. What that flag actually did was combining if we have valid eye gaze data and eye tracking was enabled by the user as active input source. 
If the demo visualizer should work as you've described there's a bug in the implementation, this code shouldn't have worked with eye tracking not being set as input source. I'm with @vaoliva  on that one - if we want this example to be fixed let's file a follow up issue and someone should thoroughly test this when doing the changes - we shouldn't mix this into this PR :)

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs (hunk 7 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs]: @vaoliva  instead of forcefully activating that flag - maybe we just want to take advantage of your patched up util function for getting the eye tracking data (InputRayUtils) and use that for determining the gameobject position - that way you don't have to interfer with what the user has set up

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs (hunk 8 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs]: IsGazeInputEyeBased is a combination of UseEyeTracking which is the user enabled setting for having eye tracking as input method and IsEyeTrackingAvailable (which i want to rename to IsEyeTrackingDataValid) which is the flag for indicating if we did receive / could process eye tracking data from the hardware.

So basically IsGazeInputEyeBased is like a IsEyeTrackingEnabledAndValid flag (maybe we should rename it to that?)

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs (hunk 6 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/DemoVisualizer/Scripts/LogStructureEyeGaze.cs]: Just to make sure this was clear, The Eye tracking demo scenes are all working after change.

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs (hunk 7 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs]: The eye tracking scenes are actually using focus info. So if gaze is not eye based scenes would be broken. Also [documentation](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/EyeTracking/EyeTracking_BasicSetup.html#enabling-eye-tracking-in-the-gazeprovider) suggests user can set up UseEyeTracking on code.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs (hunk 8 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/EyeTrackingTarget.cs]: Its more clear to me too

File: Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs (hunk 7 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.Examples/Demos/EyeTracking/General/Scripts/FollowEyeGaze.cs]: ah, i didn't see they're using the focus target as well - but then we should probably just not set that flag at all here - and just refer to the documentation - we could prompt a warning telling the user that this flag isnt' enabled but not set it ourselves. We're not doing that in any other example, so wouldn't do it in just one of our scripts.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/InputRayUtils.cs (hunk 9 above)
Comment (vaoliva) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/InputRayUtils.cs]: eyeGazeProvider.GazeOrigin / GazeDirection returns current gaze origin and direction. Using this, If gaze is head based, this util would always return head ray. What we want is a way of querying the raw eye ray, even if gaze provider is not using it, the data coming from device or simulation.

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md (hunk 11 above)
Comment (vaoliva) [reply Documentation/EyeTracking/EyeTracking_BasicSetup.md]: The info is still there. It was moved to a a new topic:

`###` Enabling eye-based gaze in the gaze provider

We are trying to separate instructions for enabling eye tracking from setup eye gaze on gaze provider. This way, user can query eye tracking data without opting for eye based gaze.

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md (hunk 12 above)
Comment (vaoliva) [reply Documentation/EyeTracking/EyeTracking_BasicSetup.md]: Same as above.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerProfileTests.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+using Microsoft.MixedReality.Toolkit.Input;
+using NUnit.Framework;
+using UnityEditor;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    class PointerProfileTests
+    {
+        private const string eyeTrackingConfigurationProfileGuid = "6615cacb3eaaa044f99b917186093aeb";
```
Comment (thalbern) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerProfileTests.cs]: nit: can you add a comment that translates that guid to the path - when kurtis replaced our hard coded paths to guids he left the paths as comments so ppl could still figure out where those assets are

File: Documentation/EyeTracking/EyeTracking_BasicSetup.md [hunk 15]
```diff
@@ -87,7 +74,18 @@ We do recommend to hide the default head gaze cursor that is attached to the MRT
     - Clone the _'DefaultMixedRealityInputPointerProfile'_ to make changes to it.
     - At the top of the _'Pointer Settings'_, you should assign an invisible cursor prefab to the _'GazeCursor'_. If you downloaded the MRTK Examples folder, you can simply reference the included _'EyeGazeCursor'_ prefab.
 
-    ![MRTK](../Images/EyeTracking/mrtk_setup_eyes_gazesettings.jpg)
+### Enabling eye-based gaze in the gaze provider
+
+In HoloLens v1, head gaze was used as primary pointing technique.
+While head gaze is still available via the _GazeProvider_ in MRTK which is attached to your [Camera](https://docs.unity3d.com/ScriptReference/Camera.html), you can check to use eye gaze instead by ticking the _'UseEyeTracking'_ checkbox in the gaze settings of the input pointer profile as shown in the screenshot below.
+
+![MRTK](../Images/EyeTracking/mrtk_setup_eyes_gazeprovider.png)
```
Comment (thalbern) [inline Documentation/EyeTracking/EyeTracking_BasicSetup.md]: nit: i know this screenshot was already there and you just adjusted it - but I would really like us to avoid having screenshots of property inspectors here - it's also mentioned in the docs guidelines - they tend to get outdated rather quickly and we can easily describe it in the text without the need to actually show it on the screenshot.

File: Assets/MRTK.Services/InputSystem/GazeProvider.cs [hunk 16]
```diff
@@ -81,13 +81,8 @@ public bool Enabled
...
         }
 
-        [SerializeField]
-        [Tooltip("If true, eye-based tracking will be used when available. Requires the 'Gaze Input' permission and device eye calibration to have been run.")]
-        [Help("When enabling eye tracking, please follow the instructions at "
-               + "https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/EyeTracking/EyeTracking_BasicSetup.html#eye-tracking-requirements "
-               + "to set up 'Gaze Input' capabilities through Visual Studio.", "", false)]
         [FormerlySerializedAs("preferEyeTracking")]
-        private bool useEyeTracking = true;
+        private bool useEyeTracking;
```
Comment (thalbern) [inline Assets/MRTK.Services/InputSystem/GazeProvider.cs]: if this used to be serialized here we should add an entry to the release notes for the next version and mark this as a breaking change so users know that they now have to change this in their profile rather than the gaze provider itself.

**Comment by thalbern** [comment]
This PR has been open for a really long time now and I think all points got addressed properly - I would suggest we do the rename change so the states are less confusing and then we'll get it in - any other concerns can be addressed in follow up PRs / please open issues for anything that should still be addressed :)

**Comment by thalbern** [comment]
> Looks good to me now - please dont' forget to document that the isEyeTrackingEnabled setting is now part of the pointer profile in our release notes as well :)

I'm not sure if the formerlyserializedas works if we're not serializing that field anymore btw, so maybe just remove it - we're mentioning this anyways in the release notes
